symbolic-unwind/src/evaluator/mod.rs

```diff
@@ -61,49 +62,40 @@ pub struct Evaluator<'memory, A, E> {
     /// operations will fail.
     memory: Option<MemoryRegion<'memory>>,
 
-    /// A map containing the values of constants.
-    ///
-    /// Trying to use a constant that is not in this map will cause evaluation to fail.
-    constants: BTreeMap<Constant, A>,
-
-    /// A map containing the values of variables.
+    /// A map containing the values of registers.
     ///
-    /// Trying to use a variable that is not in this map will cause evaluation to fail.
+    /// Trying to use a register that is not in this map will cause evaluation to fail.
     /// This map can be modified by the [`assign`](Self::assign) and
     ///  [`process`](Self::process) methods.
-    variables: BTreeMap<Variable, A>,
+    registers: BTreeMap<Register, A>,
 
     /// The endianness the evaluator uses to read data from memory.
     endian: E,
+
+    cache: BTreeMap<Expr<A>, A>,
```

Comment:
Wouldn't it be enough to just have a cache by `Register` (basically the computed value), rather than the expression? That way, you could offer an API that works like this:

1. Create a new `Evaluator`
2. Fill it with rules
3. Query for registers, either one by one or all together, which does the minimal calculations necessary.